### PR TITLE
Implement delete old AMIs feature

### PR DESCRIPTION
**Pull request by s2504s**
## What
* Implement delete old AMIs

## Why
* Old AMIs should be deleted after retention period

### CODE REVIEW

File: ami_cleanup.py [hunk 1]
```diff
@@ -18,25 +27,11 @@
...
-def lambda_handler(event, context):
-
-    reservations = ec.describe_instances(
-        Filters=[
-            {'Name': 'tag-key', 'Values': ['backup', 'Backup', 'Snapshot']},
-        ]
-    ).get(
-        'Reservations', []
-    )
+images = ec2.images.filter(Owners=[os.environ['ami_owner']],  Filters=[{'Name': 'tag-key', 'Values': ['AMIDeleteOn']}])
```
Comment (osterman) [inline ami_cleanup.py]: This does not filter on instance id. Won't there be multiple lambdas deployed at the same time? How do we keep them from stepping on each other? @comeanother

File: main.tf [hunk 2]
```diff
@@ -112,7 +120,7 @@ resource "aws_lambda_function" "ami_backup" {
 resource "aws_lambda_function" "ami_cleanup" {
   filename         = "${path.module}/ami_cleanup.zip"
   function_name    = "${module.label_cleanup.id}"
-  description      = "Cleanup old AMI backups"
+  description      = "Automatically removal AMI images of EC2 instance (remove AMI)"
```
Comment (osterman) [inline main.tf]: `Automatically remove AMIs that have expired (delete AMI)`

Comment (s2504s) [reply main.tf]: Modified

File: README.md [hunk 3]
```diff
@@ -17,8 +17,7 @@ Notes:
 
 ```
 module "lambda_ami_backup" {
-  source = "git::https://github.com/cloudposse/tf_lambda_ami_backup.git?ref=master"
-
+  source            = "git::https://github.com/cloudposse/tf_lambda_ami_backup.git?ref=master"
```
Comment (const-bon) [inline README.md]: Replace `tf_lambda_ami_backup` with `tf_ami_backup`

File: ami_cleanup.py [hunk 4]
```diff
@@ -2,13 +2,22 @@
 #
 # @author Robert Kozora <bobby@kozora.me>
 #
-# This script will search for all instances having a tag with "Backup" or "backup"
+# This script will search for all instances having a tag with "AMIDeleteOn"
```
Comment (const-bon) [inline ami_cleanup.py]: Replace `instances` with `AMIs`.

File: ami_cleanup.py [hunk 5]
```diff
@@ -2,13 +2,22 @@
...
 # on it. As soon as we have the instances list, we loop through each instance
 # and reference the AMIs of that instance. We check that the latest daily backup
 # succeeded then we store every image that's reached its DeleteOn tag's date for
 # deletion. We then loop through the AMIs, deregister them and remove all the
-# snapshots associated with that AMI.
-
+# snapshots associated withg that AMI.
```
Comment (const-bon) [inline ami_cleanup.py]: Replace `withg` with `with`.

File: ami_cleanup.py [hunk 6]
```diff
@@ -2,13 +2,22 @@
...
+
+from __future__ import print_function
+from __future__ import print_function
+from __future__ import print_function
+from __future__ import print_function
+from __future__ import print_function
+from __future__ import print_function
+from __future__ import print_function
+from __future__ import print_function
+from __future__ import print_function
```
Comment (const-bon) [inline ami_cleanup.py]: Remove line duplicates.

File: ami_cleanup.py [hunk 7]
```diff
@@ -49,60 +45,65 @@ def lambda_handler(event, context):
...
+        except IndexError:
+            deletion_date = False
+            delete_date = False
+
+        # Our other Lambda Function names its AMIs label_id-
+        # We now know these images are auto created
+        if image.name.startswith(label_id + '-' + instance_id):
+
+            # Count this image's occurance
+            imagecount = imagecount + 1
```
Comment (const-bon) [inline ami_cleanup.py]: Delete unused variable.

File: ami_cleanup.py [hunk 8]
```diff
@@ -49,60 +45,65 @@ def lambda_handler(event, context):
...
-                    backupSuccess = True
-                    print "Latest backup from " + date_fmt + " was a success"
-
-        print "instance " + instance['InstanceId'] + " has " + str(imagecount) + " AMIs"
-
-    print "============="
 
-    print "About to process the following AMIs:"
-    print imagesList
+    imagecount = 0
```
Comment (const-bon) [inline ami_cleanup.py]: Delete unused variable.

File: ami_cleanup.py [hunk 9]
```diff
@@ -49,60 +45,65 @@ def lambda_handler(event, context):
...
+                    deletion_date = [
+                        t.get('Value') for t in image.tags
+                        if t['Key'] == 'AMIDeleteOn'][0]
+                    delete_date = time.strptime(deletion_date, "%m-%d-%Y")
+            except IndexError:
+                deletion_date = False
+                delete_date = False
+
+            today_time = datetime.datetime.now().strftime('%m-%d-%Y')
+            # today_fmt = today_time.strftime('%m-%d-%Y')
```
Comment (const-bon) [inline ami_cleanup.py]: Delete unused code.

**Review by const-bon — CHANGES_REQUESTED** [review]

File: ami_cleanup.py (hunk 9 above)
Comment (s2504s) [reply ami_cleanup.py]: Deleted

File: ami_cleanup.py (hunk 7 above)
Comment (s2504s) [reply ami_cleanup.py]: Deleted

File: ami_cleanup.py (hunk 8 above)
Comment (s2504s) [reply ami_cleanup.py]: Deleted

File: ami_cleanup.py (hunk 6 above)
Comment (s2504s) [reply ami_cleanup.py]: Removed

File: ami_cleanup.py (hunk 5 above)
Comment (s2504s) [reply ami_cleanup.py]: Replaced

File: ami_cleanup.py (hunk 4 above)
Comment (s2504s) [reply ami_cleanup.py]: Replaced

**Review by osterman — APPROVED** [review]
LGTM...  I think the `AMI` prefix on `AMIDeleteOn` is redundant since the tag is applied to an AMI it does not disambiguate in any useful way.

**Comment by const-bon** [comment]
@s2504s was it tested?

**Comment by comeanother** [comment]
`ami_backup` Execution failed.

**Comment by const-bon** [comment]
Backup and Cleanup has been successfully tested manually:
1. Manually trigger `backup`.
2. Change `AMIDeleteOn` tag of newly created AMI.
3. Manually trigger `cleanup`.

**Review by const-bon — APPROVED** [review]

**Pull request merged**